lib/public_key/src/pubkey_policy_tree.erl

```diff
@@ -0,0 +1,421 @@
+-spec prune_tree(policy_tree()) -> policy_tree().
+%%
+%% Delete branches that are shorher than the total depth of the tree.
+%%--------------------------------------------------------------------
+prune_tree({} = Empty) ->
+    Empty;
+prune_tree({_, []}) ->
+    empty();
+prune_tree({Root, SubTree}) ->
+    case prune_subtree(SubTree) of
+        [] ->
+            empty();
+        NewSubTree ->
+            {Root, NewSubTree}
+    end.
+
+%%--------------------------------------------------------------------
+-spec prune_invalid_nodes(policy_tree(), [policy_node()]) -> policy_tree().
+%%
+%% Delete all invalid policy nodes and thier children.
+%%--------------------------------------------------------------------
+prune_invalid_nodes(Tree, []) ->
+    Tree;
+prune_invalid_nodes({Root, SubTree}, InvalidNodes) ->
+    case prune_invalid_nodes_subtree(SubTree, InvalidNodes) of
+        [] -> %% No leaves left the tree becomes empty
+            empty();
+        NewSubTree -> %% Keep root that is always ?anyPolicy
+            {Root, NewSubTree}
+    end.
+%%--------------------------------------------------------------------
+-spec policy_node(public_key:oid(), term(), [public_key:oid()]) -> policy_node().
+%%
+%% Creates a policy node
+%%--------------------------------------------------------------------
+policy_node(ValidPolicy, Qualifiers, ExpPolicySet) ->
+    QualifierSet = 
+        case Qualifiers of
+            asn1_NOVALUE ->
+                [];
+            _ ->
+                Qualifiers
+        end,
+    #{valid_policy => ValidPolicy,
+      qualifier_set => QualifierSet,
+      expected_policy_set => ExpPolicySet}.
+
+%%--------------------------------------------------------------------
+-spec root() -> policy_tree().
+%%
+%% Create initial root node used as start value when building the policy tree.
+%%--------------------------------------------------------------------
+root() ->
+    {any_node(), []}.
+
+%%--------------------------------------------------------------------
+-spec valid_policy_node_set(policy_tree()) -> [policy_node()].
+%%
+%% Determine the set of policy nodes whose parent nodes have a
+%% valid_policy of anyPolicy.  This is the valid_policy_node_set.
+%%--------------------------------------------------------------------
+valid_policy_node_set({#{valid_policy := ?anyPolicy}, [{_, _}| _] = SubTree}) ->
+    Parents = [Parent || {Parent, _} <- SubTree],
+    Parents ++ lists:foldl(fun(ChildNode, Acc) ->
+                                   valid_policy_node_set(ChildNode) ++ Acc
+                           end, [], SubTree);
+valid_policy_node_set({#{valid_policy := ?anyPolicy}, Leaves}) when is_list(Leaves) ->
+    Leaves;
+valid_policy_node_set(_) ->
+    [].
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+any_node() ->
```

Comment:
general comment about function name starting with `any_*`

`any_node` is ambigous in 2 ways:
- it is not clear if it is PolicyNode or TreeNode
- any_node could actually be read as Any node of any sort
- `any_policy_node` - would be slightly better I think

Reply:
sure